ldes-server-compaction/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/compaction/application/services/PaginationCompactionService.java

```diff
@@ -1,67 +1,77 @@
-		this.compactableRelationPredicate = new CompactableRelationPredicate(this.fragmentRepository);
+		this.allocationRepository = allocationRepository;
+		this.fragmentationStrategy = fragmentationStrategy;
+		this.applicationEventPublisher = applicationEventPublisher;
+		this.observationRegistry = observationRegistry;
 
 	}
 
-	public void applyCompactionStartingFromNode(Fragment fragment) {
-		Fragment firstFragmentOfCompaction = fragment;
-		while (firstFragmentOfCompactionHasOutgoingRelations(firstFragmentOfCompaction)) {
-			Optional<Fragment> secondFragmentOfCompaction = retrieveMostCompactedRelationFragment(
-					firstFragmentOfCompaction);
-			if (secondFragmentOfCompaction.isPresent()) {
-				testAndApplyPossibleCompaction(firstFragmentOfCompaction, secondFragmentOfCompaction.get());
-				firstFragmentOfCompaction = secondFragmentOfCompaction.get();
-			} else {
-				break;
-			}
-		}
-	}
+	public void applyCompactionForFragments(Set<CompactionCandidate> toBeCompactedFragments) {
+		Observation compactionObservation = Observation.createNotStarted("compaction", observationRegistry).start();
+		CompactedFragment compacted = new CompactedFragment(toBeCompactedFragments);
+
+		Fragment compactedFragment = compacted.getFragment();
+
+		fragmentRepository.saveFragment(compactedFragment);
+
+		var membersOfCompactedFragments =
+				allocationRepository.getMemberAllocationIdsByFragmentIds(compacted.getImpactedFragmentIds());
 
-	private void testAndApplyPossibleCompaction(Fragment firstFragmentOfCompaction,
-			Fragment secondFragmentOfCompaction) {
-		if (compactableRelationPredicate
-				.test(new CompactionCandidate(firstFragmentOfCompaction, secondFragmentOfCompaction))
-				&& compactableFragmentPredicate.test(secondFragmentOfCompaction)
-				&& compactableFragmentPredicate.test(firstFragmentOfCompaction)) {
-			fragmentCompactionService.compactFragments(firstFragmentOfCompaction, secondFragmentOfCompaction);
-		}
+		updateRelationsOfPredecessorFragments(compactedFragment, compacted.getFirstImpactedFragment());
+		addMembersOfFragmentsToCompactedFragment(membersOfCompactedFragments, compactedFragment, compactionObservation);
+		applicationEventPublisher.publishEvent(new FragmentsCompactedEvent(compacted.getImpactedFragmentIdentifiers()));
+		compactionObservation.stop();
 	}
 
-	private boolean firstFragmentOfCompactionHasOutgoingRelations(Fragment firstFragmentOfCompaction) {
-		return !firstFragmentOfCompaction.getRelations().isEmpty();
+	private void addMembersOfFragmentsToCompactedFragment(List<String> membersOfCompactedFragments, Fragment fragment,
+	                                                      Observation compactionObservation) {
+		membersOfCompactedFragments.forEach(memberId -> {
+			// memberModel can be null, since we explicitly use FragmentationStrategyImpl
+			fragmentationStrategy.addMemberToFragment(fragment, memberId, null, compactionObservation);
```

Comment:
We do this just to send out member allocation events. Future improvement -> bulk member allocation